### PR TITLE
feat: Added (user,state) index for talk_invitations

### LINKED ISSUE #12553: Missing index for talk_invitations database lookup
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

## Steps to reproduce
1. ???

### Expected behaviour

Efficient database operations

### Actual behaviour

```sql
SELECT
  COUNT (*)
FROM
  `oc_talk_invitations`
WHERE
  (`user_id` = ?)
  AND (`state` = ?)
```

not using an index.

Looking at https://github.com/nextcloud/spreed/blob/38bfcad79113a110084789db21e93cbaed668627/lib/Model/InvitationMapper.php#L71-L100 it seems like an index on (user_id, state) would be a good candidate.

## Talk app

**Talk app version:** (see apps administration page: `/index.php/settings/apps`)

**Custom Signaling server configured:** yes/no and version (see  Talk administration settings: `/index.php/settings/admin/talk#signaling_server`)

**Custom TURN server configured:** yes/no (see Talk administration settings: `/index.php/settings/admin/talk#turn_server`)

**Custom STUN server configured:** yes/no (see  Talk administration settings: `/index.php/settings/admin/talk#stun_server`)

## Browser

**Microphone available:** yes/no

**Camera available:** yes/no

**Operating system:** Windows/Ubuntu/Mac/...

**Browser name:** Firefox/Chrome/Safari/...

**Browser version:** 124/125/...

### Browser log

<details>
```
Insert your browser log here, this could for example include:
a) The javascript console log
b) The network log
c) ...
```

</details>

## Server configuration

**Operating system**: Ubuntu/RedHat/...

**Web server:** Apache/Nginx

**Database:** MySQL/Maria/SQLite/PostgreSQL

**PHP version:** 8.1/8.2/8.3

**Nextcloud Version:** (see administration page)

**List of activated apps:**

<details>

```
If you have access to your command line run e.g.:
sudo -u www-data php occ app:list
from within your server installation folder
```
</details>

**Nextcloud configuration:**

<details>

```
If you have access to your command line run e.g.:
sudo -u www-data php occ config:list system
from within your Nextcloud installation folder
```
</details>

### Server log (data/nextcloud.log)
<details>

```
Insert your server log here
```
</details>

Query ID 805953543f46cc758e6b0239a23b8bfd

### PR DESCRIPTION
**Pull request by skalidindi53**
### ☑️ Resolves

* Fix #12553 

### 🚧 Tasks

- [x] Added the user,state index for talk_invitations database table

### 🏁 Checklist

- [x] ⛑️ Tests (unit and/or integration) are included or not possible
- [x] 📘 API documentation in `docs/` has been updated or is not required
- [x] 🔖 Capability is added or not needed

### CODE REVIEW

**Comment by nickvergessen** [comment]
Migrations are only ever executed once. So you need to create a new migration file (using the `occ migration:generate` command)

File: lib/Migration/Version20000Date20240716190335.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	 * @param array $options
+	 * @return null|ISchemaWrapper
+	 */
+	public function changeSchema(IOutput $output, Closure $schemaClosure, array $options): ?ISchemaWrapper {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+
+		$table = $schema->getTable('talk_invitations');
+		if (!$table->hasIndex('talk_user_state_idx')) {
+			$table->addIndex(['user_id', 'state'], 'talk_user_state_idx');
```
Comment (nickvergessen) [inline lib/Migration/Version20000Date20240716190335.php:30]: Not a problem, but just so you know. PostgreSQL is automatically adding an IDX suffix. Oracle has another one. So we just skip them usually :D
```suggestion
		if (!$table->hasIndex('talk_user_state')) {
			$table->addIndex(['user_id', 'state'], 'talk_user_state');
```

Comment (skalidindi53) [reply lib/Migration/Version20000Date20240716190335.php]: Did not know that, thanks for letting me know! Just updated the name

File: lib/Migration/Version20000Date20240716190335.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * SPDX-FileCopyrightText: 2023 Nextcloud GmbH and Nextcloud contributors
```
Comment (nickvergessen) [inline lib/Migration/Version20000Date20240716190335.php:6]:
```suggestion
 * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
```

**Review by nickvergessen — APPROVED** [review]

**Review by nickvergessen — APPROVED** [review]

**Pull request merged**